File: p216m/combination_sum3.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function (k, n) {
  let ans = [];
  /**
   * Recursive Backtrack helper function
   * closure {number[][]} solutions: output
   * @param {number[]} workPath:  temporary working path
   * @param {number}   remTarget: ramaining target
   * @param {number}   start: index in candidates to start search
   */
  let recBacktracing = function (workPath, remTarget, start) {
    if (workPath.length == k) {
      if (remTarget === 0) ans.push(Array.from(workPath));
      return;
    }
    for (let i = start; i <= 9; ++i) {
      workPath.push(i);
      recBacktracing(workPath, remTarget - i, i + 1);
      workPath.pop();
    }
  };
  recBacktracing([], n, 1);
  return ans;
};
[
  [0, 0],
  [1, 9],
  [3, 7],
  [3, 9],
].forEach(([k, n]) => {
  console.log(
    `Combinations of ${k} digits summing to ${n} ->`,
    combinationSum3(k, n)
  );
});

File: p216m/combination_sum3.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> list[list[int]]:
        ans = []

        def backtrack(path: list[int], k: int, rem: int, start: int) -> None:
            if k == 0:
                if rem == 0:
                    ans.append(path)
                return

            for num in range(start, 10):
                backtrack(path + [num], k - 1, rem - num, num + 1)

        backtrack([], k, n, 1)
        return ans


# TESTS
for k, n, expected in [
    (3, 7, [[1, 2, 4]]),
    (3, 9, [[1, 2, 6], [1, 3, 5], [2, 3, 4]]),
    (4, 1, []),
]:
    sol = Solution()
    actual = sol.combinationSum3(k, n)
    print("Combination sum k =", k, ", n =", n, "->", actual)
    assert actual == expected
